### PR TITLE
Run clang-tidy in GitHub CI

### PR DESCRIPTION
**Pull request by masterleinad**
Alternative to #14589. Maybe, we can at least use this as a temporary solution.

### CODE REVIEW

**Comment by masterleinad** [comment]
This uses `CMake 3.25.1`.

**Review by tamiko — APPROVED** [review]
I like this. Thanks for putting the github workflow together!

File: source/lac/trilinos_sparsity_pattern.cc [hunk 1]
```diff
@@ -117,7 +117,7 @@ namespace TrilinosWrappers
 
 
   SparsityPattern::SparsityPattern(SparsityPattern &&other) noexcept
-    : SparsityPatternBase(other)
+    : SparsityPatternBase(std::move(other))
```
Comment (bangerth) [inline source/lac/trilinos_sparsity_pattern.cc:120]: Oh yes!

**Review by bangerth — COMMENTED** [review]
The actual CI file is outside my realm of expertise, but the remainder of the changes are fine.

**Comment by bangerth** [comment]
In case this doesn't get merged quickly: If you wanted to break out the code changes, these are uncontroversial and can be merged separately.

**Comment by masterleinad** [comment]
In comparison to our current (or updated) `clag-tidy` approach
- this enables more dependencies (we could even enable everything we use for the docker image).
- this uses `clang-12` instead of `clang-14` since that's what we natively get on `Ubuntu 22.04` (shouldn't be too hard to change though if necessary).
- this runs much longer ~40m vs ~7m but that's in line with the other GitHub CI runs.

**Comment by tjhei** [comment]
> In comparison to our current (or updated) `clag-tidy` approach

I am fine to get rid of the Jenkins version if the Github runner is acceptable...

**Comment by tamiko** [comment]
@masterleinad @tjhei I have no strong feelings either way (whether Github CI or Jenkins) 

I think the strongest point to make is the much faster turnaround that Jenkins provides. I can see that bindly trying to please a `clang-tidy` instance with a 40 min turnaround might be frustrating...

Maybe we should add a clang-tidy target similar to our indent target? :-)

In either case - I think that the functional changes in this pull request (switch to clang-12 with a manageable amount of fallout that is already addressed in this PR).

I would like to settle on a solution soon - currently all of the CMake refactoring is stalled.

**Comment by tjhei** [comment]
I can also run clang-12 instead (or we just disable anything new)...

**Comment by masterleinad** [comment]
I don't have strong feelings either way either. I just saw some value in trying with GitHub to remove the roadblock given that we need all kinds of workarounds for #14589 and we still don't quite understand what the current problem is.
Maybe, we just do this as a short-term solution while trying to fix #14589 so that we get better turnaround times.

**Review by drwells — APPROVED** [review]

**Pull request merged**